Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,190 @@
+# Objective 
+The objective of this document is to provide details on how to build a generic queue transport inside CoreWCF so that it can enable various queue binding. This document is based on the discussion outlined here https://github.com/CoreWCF/CoreWCF/issues/164 .
+
+# Why are we building
+When WCF was build there were not many queing technique and there was no concept of cloud. With multiple cloud providers supporting queing technique and many queing technique widely available, we wanted to build a generic transport, so that various queuing mechanism can be supported in CoreWCF.
+
+# Class Diagram
+![CoreWCF Generic Queue](/Documentation/DesignDocs/generic_queue_class_diagram.png?raw=true)
+
+# Class Details
+#### IQueueMiddlewareBuilder and QueueMiddlewareBuilder
+- IQueueMiddlewareBuilder provides signatures for 2 important methods, ```Use``` and ```Build```. 
+```csharp
+internal interface IQueueMiddlewareBuilder 
+{ 
+   IServiceProvider QueueServices { get; set; } 
+   IDictionary<string, object> Properties { get; } 
+   IQueueMiddlewareBuilder Use(Func<QueueMessageDispatcherDelegate, QueueMessageDispatcherDelegate> middleware); 
+   QueueMessageDispatcherDelegate Build(); 
+} 
+```
+- ```Use``` adds middleware delegates to the Queue pipeline 
+- ```Build``` builds the delegate used by this application to process requests received from the queue.
+#### QueueMiddleWareBuilderExtension
+- QueueMiddleWareBuilderExtension defines ```UseMiddleware``` method which specifically search for ```InvokeAsync``` method in each middleware class and construct it as delegate to be invoked by the middleware pipeline. This is similar to the NetTcp class UseMiddlewareFramingConnectionHandshakeExtensions and the ASP.NET Core UseMiddlewareExtensions..
+```csharp
+public class QueueMiddleWareBuilderExtension
+{
+    internal const string InvokeAsyncMethodName = "InvokeAsync";
+    public static IQueueMiddlewareBuilder UseMiddleware<TMiddleware>(this IQueueMiddlewareBuilder app, params object[] args)
+    {
+         return app.UseMiddleware(typeof(TMiddleware), args);
+    }
+    
+    public static IQueueMiddlewareBuilder UseMiddleware(this IQueueMiddlewareBuilder app, Type middleware, params object[] args)
+    {
+        //TODO
+     }
+    }
+```
+#### QueueServiceWebHostBuilderExtension
+- QueueServiceWebHostBuilderExtension is the WebHostBuilder extension which stiches all the middleware pipeline and ```QueuePollingService``` which will be called during service startup for queuing mechanism to work. e:g:- 
```

Comment:
Make sure any specified types are quoted as inline code. Eq `QueueServiceWebHostBuilderExtension` as opposed to QueueServiceWebHostBuilderExtension

Reply:
You should provide the class design here. Also the example should be updated to .NET 6 style syntax. See the sample app [here](https://github.com/CoreWCF/samples/blob/main/Scenarios/Logging-and-Dependency-Injection/LoggingSample/Program.cs) for an example.

Reply:
We don't need to add CoreWCF into the method name. It's not `UseAspNetCoreKestrel`. We should just be bold and call it `UseQueues`. This is also the wrong place to call the method, as Use verbs are applied to builders, Add verbs to collections. I've been trying to work out where we should apply the `UseQueues` method and I'm beginning to think for CoreWCF 2.0 we need to take a bit of a step back to look at the api's and how you configure CoreWCF in the future. As of right now we have NetTcp (which uses UseConnectionHandler and is based on Kestrel) and HTTP based transports, so everything builds off of a WebHost. We're about to add queue's, and then we have named pipe's coming up, neither of which will build on top of Kestrel. Then there's TCP Port Sharing, and the IIS hosted model in the future where the IIS Integration IServer type is used, both of which will have us using sockets outside of an IServer. There's also been some demand for UDP based bindings too. There's a new model in .NET 6 where a Web based application (mean HTTP based) is just another IHostedService implementation. I think we need to start thinking about configuring CoreWCF transports outside of the IWebHostBuilder world (as you might not have an HTTP endpoint) and have some mechanism build around `IHostBuilder`. I don't want to hold up merging this based on that (as it will take some designing), this is more of a heads up that this portion is likely to evolve. It shouldn't affect the rest of the design as IHostedService is still the right way to go. Basically everything should become an IHostedService and this is around how do we add/configure the various transports in the host. We also need to solve the problem of not wanting to configure your services and endpoints in multiple places.

Reply:
Agree, this will evolve as we build more and more layers..